apps/sms/js/thread_list_ui.js

```diff
@@ -282,6 +286,71 @@ var ThreadListUI = {
         }
 
         break;
+      case 'contextmenu':
+        if (this.inEditMode) {
+          return;
+        }
+
+        parent = event.target.parentNode;
+        parentThreadId = parent.dataset.threadId;
+        var thread = Threads.get(parentThreadId);
+
+        if (parentThreadId) {
+          event.preventDefault();
+          event.stopPropagation();
+          // Show options per single thread
+          var params = {
+            type: 'action',
+            header: { l10nId: 'thread-options' },
+            items:[]
+          };
+          threadIdToDelete.push(parentThreadId);
+
+          params.items.push(
+            {
+              l10nId: 'delete-thread',
+              method: function deleteThread(threadIdToDelete) {
+                ThreadListUI.performDeletion(threadIdToDelete);
+              },
+              params: [threadIdToDelete]
+            }
+          );
+
+          if(typeof Threads.get(+parentThreadId) !== 'undefined') {
+            var isRead = thread && (thread.unreadCount > 0);
+
+            isRead ? l10nKey = 'mark-as-read' : l10nKey = 'mark-as-unread';
+
+            params.items.push(
+              {
+                l10nId: l10nKey,
+                method: function markReadUnread(parentThreadId) {
+                  {
+                    var selectThread =
+                      document.getElementById('thread-' + thread.id);
+                    thread.unreadCount = isRead ? 0 : 1;
+                    if (isRead) {
+                      selectThread.classList.remove('unread');
+                    } else {
+                      selectThread.classList.add('unread');
+                    }
+                    MessageManager.markThreadRead(+parentThreadId, isRead);
+                  }
+                },
+                params: [+parentThreadId]
+              }
```

Comment:
I'm sure you can reuse the existing markReadUnread :) Possibly splitting it just as you did for `delete` ?